### ISSUE TITLE
Host's SSH agent not forwarded to development container on macOS

### ISSUE THREAD

**Issue by lsorber** [microsoft/vscode-remote-release#4024]
### Versions

- VSCode Version: 1.51.1
- Local OS Version: macOS 10.15.7
- Remote OS Version: Debian 10
- Remote Extension/Connection Type: Docker

### Bug description

It appears that the SSH agent is not forwarded correctly to the development container on macOS. Cloning a private git repository on the host over SSH works as expected, but the same git clone within the development container fails:
```
❯ git clone git@github.com:lsorber/my-private-repo.git  # In dev container
Cloning into 'my-private-repo'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

### Steps to Reproduce

1. Create a development container (`devcontainer.json` spec below).
2. Try to clone a private git repository over SSH.

<details>
<summary>devcontainer.json spec:</summary>

```
{
    "name": "entities",
    "image": "continuumio/miniconda3:latest",
    "extensions": [
        "ms-python.python",
        "ms-python.vscode-pylance",
        "ms-toolsai.jupyter",
        "eamodio.gitlens"
    ],
    "settings": {
        "terminal.integrated.shell.linux": "/usr/bin/fish",
        "python.pythonPath": "/opt/conda/envs/entities-env/bin/python",
        "python.terminal.activateEnvironment": false,
        "python.terminal.activateEnvInCurrentTerminal": false
    },
    "postCreateCommand": "./tasks/init.sh devcontainer"  // Installs fish shell and creates conda environment
}
```
</details>

### Workaround

Explicitly forwarding the SSH agent in `devcontainer.json` with the snippet below resolves the issue, but is specific to macOS:
```
    "mounts": [
        "source=/run/host-services/ssh-auth.sock,target=/run/host-services/ssh-auth.sock,type=bind,consistency=cached"
    ],
    "containerEnv": {
        "SSH_AUTH_SOCK": "/run/host-services/ssh-auth.sock"
    },
```

**Comment by chrmarti** [comment]
Could you attach the log (`F1` > `Remote-Containers: Show log`) from when you don't use your workaround?

**Comment by lsorber** [comment]
Sure, here are the logs when I start the dev container:

<details>

<summary>Remote-Containers: Show Log (workaround disabled)</summary>

```
[4 ms] Remote-Containers 0.148.1 in VS Code 1.51.1 (e5a624b788d92b8d34d1392e4c4d9789406efe8f).
[29 ms] Start: Resolving remote
[32 ms] Setting up container for folder or workspace: /Users/.../entities

[34 ms] Start: Check Docker is running
[34 ms] Start: Run: docker info
[817 ms] Start: Run: git rev-parse --show-cdup
[829 ms] Start: Run: docker ps -q -a --filter label=vsch.local.folder=/Users/.../entities --filter label=vsch.quality=stable
[1045 ms] Start: Run: docker inspect --type container 160a6f406050
[1286 ms] Start: Inspecting container
[1286 ms] Start: Run: docker inspect --type container 160a6f406050a86f2e6c69db7d17b774c4395a31246189c324bb9e9eb1f7ec48
[1531 ms] Start: Run: docker exec -i -u root -e VSCODE_REMOTE_CONTAINERS_SESSION=889ded31-1855-468d-a427-9838c4e17fd61606393390692 160a6f406050a86f2e6c69db7d17b774c4395a31246189c324bb9e9eb1f7ec48 /bin/sh
[1536 ms] Start: Run in container: uname -m
[1897 ms] x86_64
[1897 ms] 
[1898 ms] Start: Run in container: (cat /etc/os-release || cat /usr/lib/os-release) 2>/dev/null
[1916 ms] PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
[1916 ms] 
[1917 ms] Start: Run in container: cat /etc/passwd
[1923 ms] Start: Setup shutdown monitor
[1927 ms] Forking shutdown monitor: /Users/laurent/.vscode/extensions/ms-vscode-remote.remote-containers-0.148.1/dist/shutdownMonitorProcess /var/folders/vj/lqch8d4x4gd_t_3fd64lnttr0000gn/T/vscode-remote-containers-e68c72e300fdc090fd158c1a3ae9c53337a5bfdb.sock singleContainer Info /Users/laurent/Library/Application Support/Code/logs/20201118T170645/exthost5/ms-vscode-remote.remote-containers 1606393391839
[1931 ms] Start: Run in container: test -d /root/.vscode-server
[1936 ms] 
[1936 ms] 
[1936 ms] Start: Run in container: set -o noclobber ; mkdir -p '/root/.vscode-server/data/Machine' && { > '/root/.vscode-server/data/Machine/.writeMachineSettingsMarker' ; } 2> /dev/null
[1945 ms] 
[1945 ms] 
[1945 ms] Exit code 2
[1946 ms] Start: Run in container: test -d /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f
[1949 ms] 
[1949 ms] 
[1949 ms] Start: Launching Remote-Containers helper.
[1950 ms] Start: Run: gpgconf --list-dir agent-extra-socket
[1958 ms] Start: Run in container: cat <<'EOF-/tmp/vscode-remote-containers-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js' >/tmp/vscode-remote-containers-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js
[1966 ms] 
[1966 ms] 
[1966 ms] Start: Run in container: # Test for /root/.ssh/known_hosts and ssh
[1972 ms] /root/.ssh/known_hosts exists
[1973 ms] 
[1973 ms] Exit code 1
[1973 ms] Start: Run in container: cat <<'EOF-/tmp/vscode-remote-containers-server-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js' >/tmp/vscode-remote-containers-server-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js
[1980 ms] 
[1980 ms] 
[1980 ms] Start: Run in container: command -v git >/dev/null 2>&1 && git config --system credential.helper '!f() { /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/node /tmp/vscode-remote-containers-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js $*; }; f' || true
[1981 ms] Start: Run: docker exec -i -u root -e REMOTE_CONTAINERS_SOCKETS=["/tmp/vscode-ssh-auth-24a873ea7fae7dd8c2599e9c573c29234c5030fb.sock"] -e REMOTE_CONTAINERS_IPC=/tmp/vscode-remote-containers-ipc-24a873ea7fae7dd8c2599e9c573c29234c5030fb.sock 160a6f406050a86f2e6c69db7d17b774c4395a31246189c324bb9e9eb1f7ec48 /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/node /tmp/vscode-remote-containers-server-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js
[1993 ms] 
[1993 ms] 
[1993 ms] Start: Run in container: set -o noclobber ; mkdir -p '/root/.vscode-server/data/Machine' && { > '/root/.vscode-server/data/Machine/.installExtensionsMarker' ; } 2> /dev/null
[2006 ms] 
[2006 ms] 
[2006 ms] Exit code 2
[2008 ms] Start: Run in container: set -o noclobber ; mkdir -p '/root/.vscode-server/data/Machine' && { > '/root/.vscode-server/data/Machine/.postCreateCommandMarker' ; } 2> /dev/null
[2018 ms] 
[2018 ms] 
[2018 ms] Exit code 2
[2021 ms] Start: Run in container: # Test for /root/.gitconfig and git
[2026 ms] /root/.gitconfig exists
[2026 ms] 
[2026 ms] Exit code 1
[2026 ms] Start: Run in container: command -v git >/dev/null 2>&1 && git config --global credential.helper '!f() { /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/node /tmp/vscode-remote-containers-24a873ea7fae7dd8c2599e9c573c29234c5030fb.js $*; }; f' || true
[2037 ms] 
[2037 ms] 
[2038 ms] Start: Run in container: for pid in `cd /proc && ls -d [0-9]*`; do { echo $pid ; readlink -f /proc/$pid/cwd ; xargs -0 < /proc/$pid/environ ; xargs -0 < /proc/$pid/cmdline ; } ; echo ; done 2>/dev/null
[2725 ms] Extension host agent is already running.
[2725 ms] Start: Run in container: cat /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/.devport 2>/dev/null
[2735 ms] 35227
[2735 ms] 
[2736 ms] Forwarding local port 35227 to container port 35227
[2748 ms] Start: Run: docker exec -i -u root -e VSCODE_REMOTE_CONTAINERS_SESSION=889ded31-1855-468d-a427-9838c4e17fd61606393390692 160a6f406050a86f2e6c69db7d17b774c4395a31246189c324bb9e9eb1f7ec48 /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/node -e 
[2753 ms] Start: Run: docker exec -i -u root -e VSCODE_REMOTE_CONTAINERS_SESSION=889ded31-1855-468d-a427-9838c4e17fd61606393390692 160a6f406050a86f2e6c69db7d17b774c4395a31246189c324bb9e9eb1f7ec48 /root/.vscode-server/bin/e5a624b788d92b8d34d1392e4c4d9789406efe8f/node -e 
[4611 ms] Start: Run in container: cat /proc/3938/environ
```

</details>

**Comment by chrmarti** [comment]
It looks like the SSH socket is being set up. Do you have 'Inherit Env' in the settings enabled? That is needed to get SSH_AUTH_SOCK to the terminal.
![image](https://user-images.githubusercontent.com/9205389/101048166-b5f9ce00-3582-11eb-9a15-f3ca96b9cfd2.png)

Does `ssh-add -l` work?

**Comment by lsorber** [comment]
Hi @chrmarti:

- Yes, `Terminal > Integrated: Inherit Env` is True in the User, Remote Dev Container, and Workspace settings.
- The output of `ssh-add -l` in the Dev Container is `The agent has no identities.`

Anything else I can do to debug this?

**Comment by chrmarti** [comment]
What is the output of `ssh-add -l` on the host machine?
- If that has no identities, you need to add them with `ssh-add <private-key-file>`. Then check if it works.
- If that has identities, what is the value of `SSH_AUTH_SOCK` in the container?

**Comment by lsorber** [comment]
Thanks for the pointers @chrmarti. I investigated a bit further and I found that `ssh-add -l`'s output on macOS depends on a few factors:

1. For context, my SSH keys are added to the SSH agent automatically through `~/.ssh/config`, using the same approach that GitHub recommends [1].
2. After a fresh reboot, `ssh-add -l` outputs `The agent has no identities.` on the host.
3. However, after running `ssh -T git@github.com` or cloning a private GitHub repo, the SSH keys are _lazily_ added to the SSH agent, and the output of `ssh-add -l` then shows that that SSH key is added.

In conclusion: lazily adding SSH keys to the SSH agent works fine for most tasks on the host, but it also means that if you haven't (lazily or explicitly) added your SSH keys after rebooting, the Dev Container won't have access to them either.

Some possible solutions:
1. Workaround: explicitly add all SSH keys to the SSH agent after login with `ssh-add -A` [2]. The downside is that this seems to be an extra step for anyone wanting to use Dev Containers.
2. Solution: on macOS, have the Remote - Containers extension run `ssh-add -A` when launching a Dev Container to make sure that all keys are added to the SSH agent before starting the container.

[1] https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#adding-your-ssh-key-to-the-ssh-agent
[2] https://github.com/jirsbek/SSH-keys-in-macOS-Sierra-keychain#solution-2

**Comment by chrmarti** [comment]
Thanks for investigating! The documentation for AddKeysToAgent suggests that there is a way to make it ask the user for confirmation. We wouldn't want to trigger that automatically every time because we don't know if the user is actually going to use the SSH keys in the current session.

**Comment by lsorber** [comment]
@chrmarti a solution could be to change the extension's method of forwarding the SSH agent to the one described in the OP by mounting `SSH_AUTH_SOCK=/run/host-services/ssh-auth.sock`. That would allow the guest to lazily load SSH keys the same way as the host does.

If that's not an option, I think the docs on this topic [1] should be updated to recommend macOS users to run `ssh-add -A` before starting the Dev Container.

[1] https://code.visualstudio.com/docs/remote/containers#_sharing-git-credentials-with-your-container

**Comment by floatingpurr** [comment]
Interesting topic. Just arrived here for the same problem. Still looking for a clean solution to use the host's ssh keys in the devcontainer with no hassle!

**Comment by chrmarti** [comment]
We have avoided mounting the SSH socket because that doesn't work when the Docker daemon is running on another machine.

I think running `ssh-add -A` automatically might be unexpected. E.g., a user might run `ssh-add -A`, do some SSH work and then run `ssh-add -D` (removing all keys from the agent) to minimize the time the keys are available in the agent for security reasons.

We could add it as a feature that is disabled by default and can be enabled with a user setting (might be hard to discover). Or maybe there is a way to inform the user that the agent has no keys at the time the first connection to the agent is established.

**Comment by floatingpurr** [comment]
Got it! It would be great if this feature were available out of the box. I understand that there are several security drawbacks, those can be acceptable for some use cases, though.

**Comment by floatingpurr** [comment]
> @chrmarti a solution could be to change the extension's method of forwarding the SSH agent to the one described in the OP by mounting `SSH_AUTH_SOCK=/run/host-services/ssh-auth.sock`. That would allow the guest to lazily load SSH keys the same way as the host does.
> 
> If that's not an option, I think the docs on this topic [1] should be updated to recommend macOS users to run `ssh-add -A` before starting the Dev Container.
> 
> [1] https://code.visualstudio.com/docs/remote/containers#_sharing-git-credentials-with-your-container

Is there a way to run `ssh-add ...` automatically each time the devcontainer runs?

**Comment by lsorber** [comment]
@floatingpurr I tried adding `"initializeCommand": "ssh-add -A || true"` to `devcontainer.json` [1], but it does not seem to solve the issue. The only workaround that does work that I know of is the one I posted in the OP.

[1] https://code.visualstudio.com/docs/remote/devcontainerjson-reference

**Comment by floatingpurr** [comment]
I'm on macOS 10.15.7 and VSCode: 1.52.1 too but the OP workaround doesn't work in my case.

Surprisingly enough, if I add `"initializeCommand": "ssh-add path/to/my/rsa"`, it works! (_Caveat: The commands execute from the workspaceFolder locally, therefore pay attention to the path!_)

You might want NOT to hardcode the `path/to/my/rsa` and other init stuff, especially if you check `devcontainer.json` in your codebase for dev purposes and you do not want to pollute the dev env with your specific settings. In my case, I use `"initializeCommand" : ".devcontainer/init.sh || true"`. My `init.sh` looks like this:

```bash
#!/usr/bin/env bash

# Commands run on the local machine before the container is created

ssh-add ../../.ssh/id_rsa
```

... and I added it to `.gitignore`, therefore each dev can create her/his super-specific `init.sh` and add there custom init stuff.

Still not convinced that it is a good way to manage ssh forwarding. Maybe VSCode is going to improve its features related to this kind of things in the near future.

**Comment by floatingpurr** [comment]
Any updates about this feature? :)

**Comment by 0x6A75616E** [comment]
Not sure if this is directly related to this issue, but in macOS you have to give VSCode access to the Full Disk from `Settings` > `Security & Privacy` so that it can access the sock file.

**Comment by kelvin1794** [comment]
What should be the workaround, currently? 

I'm having the same issue and have tried everything that was suggested above but none works...

**Comment by TheBestMoshe** [comment]
`"initializeCommand": "ssh-add -A ` did not work for me. When running `ssh-add -A` on my host system I get
```
No identity found in the keychain.
```

This worked did work:
```
"initializeCommand": "find ~/.ssh/ -type f -exec grep -l 'PRIVATE' {} \\; | xargs ssh-add",
```

**Comment by biltongza** [comment]
I am getting an issue where the ssh-agent on my WSL2 host says there are identities added, but the devcontainer says no identities are added to the agent.

![image](https://user-images.githubusercontent.com/2845704/132089093-1ba755cc-61a6-4040-a6c3-52df516c3270.png)

**Comment by chrmarti** [comment]
@biltongza This looks like a different problem. Could you open a new issue and include the log from when you are connected to the container (`F1` > `Remote-Containers: Show Container Log`)?

**Comment by jginss** [comment]
I'm currently banging my head with the same @biltongza and @chrmarti, do you have a link to the new issue ?

**Comment by vadixidav** [comment]
This still seems to be an issue.

**Comment by pbasimfa** [comment]
I am having the same issue with the devcontainer, but my issue  is not resolved even when using `"initializeCommand": "ssh-add ../../.ssh/id_rsa"`. `ssh-add -l` returns "Could not open a connection to your authentication agent." after adding the above command to `devcontainer.json`. When I run the ssh agent in the background and `ssh-add`, I will get "The agent has no identities.", after running `ssh-add -l` again. I have added the "remote ssh" and "remote containers" extensions in the `devcontainer.json` extensions. My ssh config file is configured according to github guidelines.

**Comment by pbasimfa** [comment]
> I am having the same issue with the devcontainer, but my issue is not resolved even when using `"initializeCommand": "ssh-add ../../.ssh/id_rsa"`. `ssh-add -l` returns "Could not open a connection to your authentication agent." after adding the above command to `devcontainer.json`. When I run the ssh agent in the background and `ssh-add`, I will get "The agent has no identities.", after running `ssh-add -l` again. I have added the "remote ssh" and "remote containers" extensions in the `devcontainer.json` extensions. My ssh config file is configured according to github guidelines.

In my case the solution was to add this to the `devcontainer.json` (tested for Github):
	`"containerEnv": {
		"GITHUB_TOKEN": "${localEnv:GITHUB_TOKEN}",
		"GITHUB_USER": "${localEnv:GITHUB_USER}"
	},`

**Comment by jamie-sgro** [comment]
For what it's worth I was able to get a devcontainer working with installed private repositories by using `"initializeCommand"` to first build my dev container like so:
```bash
DOCKER_BUILDKIT=1 docker build --ssh default -t my-image:latest <path to container>
```
The --ssh default wont work for windows out of the box but you could manually specify it. More info [here](https://docs.docker.com/develop/develop-images/build_enhancements/#using-ssh-to-access-private-data-in-builds).

In your Dockerfile, I also needed prepend your git clone with the `--mount=type=ssh`.

In the docker-compose that I route to from the `devcontainer.json` I swapped from a build to using a pre-built image:

```diff
services:
  app:
-    build:
-      context: ..
-      dockerfile: .devcontainer/Dockerfile
-      args:
-        # Update 'VARIANT' to pick a version of Python: 3, 3.10, 3.9, 3.8, 3.7, 3.6
-        # Append -bullseye or -buster to pin to an OS version.
-        # Use -bullseye variants on local arm64/Apple Silicon.
-        VARIANT: "3.9"
-        # Optional Node.js version to install
-        NODE_VERSION: "none"
+    image: my-image:latest
```

**Comment by robwilkes** [comment]
I'm not using extension Remote-Container, I'm using Remote-SSH with Docker extension installed on remote. I'm able to right-click and Build a Dockerfile which clones a private repository using `RUN --mount=type=ssh` by adding the following to  vscode settings.json:

"docker.commands.build": "DOCKER_BUILDKIT=1 eval \`ssh-agent\` && ssh-add ~/.ssh/id_rsa && docker build --ssh default --pull --rm -f \"${dockerfile}\" -t ${tag} \"${context}\" && ssh-agent -k"

**Comment by ahmadrio** [comment]
> `"initializeCommand": "ssh-add -A ` did not work for me. When running `ssh-add -A` on my host system I get
> 
> ```
> No identity found in the keychain.
> ```
> 
> This worked did work:
> 
> ```
> "initializeCommand": "find ~/.ssh/ -type f -exec grep -l 'PRIVATE' {} \\; | xargs ssh-add",
> ```

thanks it works 👍

**Comment by dslatkin** [comment]
I'm not sure if my issue is related, but the first time I build and connect to a dev container, I'm not able to clone a repo using SSH from my host machine. I get the same error as others trying to clone a private repo.

If I run `env` I can see

```
SSH_AUTH_SOCK=/tmp/ssh-XXXXXXVqcT3M/agent.636
```

when it connects the first time right after building. If I disconnect from the container and reconnect (and all future reconnects), I get a variable like

```
SSH_AUTH_SOCK=/tmp/vscode-ssh-auth-1d31887b20b2c25bedf0e097a3e0c1ea8f215d88.sock
```

and I can clone the repo and do everything as needed with SSH successfully.

**Comment by jhonnyoliveira** [comment]
> `"initializeCommand": "ssh-add -A ` did not work for me. When running `ssh-add -A` on my host system I get
> 
> ```
> No identity found in the keychain.
> ```
> 
> This worked did work:
> 
> ```
> "initializeCommand": "find ~/.ssh/ -type f -exec grep -l 'PRIVATE' {} \\; | xargs ssh-add",
> ```

Thank you, it was the only way that worked for me.
